### ISSUE TITLE
1.6.1 breaks oauth/openid token request

### ISSUE THREAD

**Issue by bilias** [roundcube/roundcubemail#8878]
After upgrading to 1.6.1 I cannot login with keycloak/openid
I rolled back and everything works again.

The error message says about "Incorrect redirect_uri"
My redirect URI is https://webmail.example.com/index.php/login/oauth
and this is what roundcube sends to keycloak.

In roundcube error log I get:
[24-Jan-2023 09:42:47 +0200]: <95d4lk4k> PHP Error: OAuth token request failed: Client error: `POST https://keycloak.example.com/realms/example/protocol/openid-connect/token` resulted in a `400 Bad Request` response:
{"error":"invalid_grant","error_description":"Incorrect redirect_uri"}
; webmail.example.com GuzzleHttp/7 - [24/Jan/2023:09:42:47 +0200] "POST /realms/example/protocol/openid-connect/token HTTP/1.1" 400 70 in /var/www/html/webmail/web/program/include/rcmail_oauth.php on line 317 (GET /index.php/login/oauth?state=zqmEF8Fb8XGN&session_state=2b1586b7-3736-44ff-b4bd-413226eedf15&code=dfe0c144-7071-44b2-a301-e5fd9265595f.2b1586b7-3736-44ff-b4bd-413226eedf15.ce59614c-9d5a-4b2b-877d-e1b6e07d142c)

On keycloak side I get:

2023-01-24 09:42:47,349 INFO  [org.keycloak.events] (executor-thread-27) type=LOGIN, realmId=fb3808fd-64a6-41b2-acd5-010db67ee826, clientId=webmail-example-com userId=877ef0ab-a97a-454c-be24-1af112fefaf4, ipAddress=xx.xx.xx.xx, auth_method=openid-connect, auth_type=code, response_type=code, redirect_uri=https://webmail.example.com/index.php/login/oauth, consent=no_consent_required, code_id=2b1586b7-3736-44ff-b4bd-413226eedf15, response_mode=query, username=bilias, authSessionParentId=2b1586b7-3736-44ff-b4bd-413226eedf15, authSessionTabId=njKKy_6pMLE

2023-01-24 09:42:47,493 ERROR [org.keycloak.events] (executor-thread-27) type=CODE_TO_TOKEN_ERROR, realmId=fb3808fd-64a6-41b2-acd5-010db67ee826, clientId=webmail-example-com, userId=877ef0ab-a97a-454c-be24-1af112fefaf4, ipAddress=xx.xx.xx.xx, error=invalid_code, grant_type=authorization_code, code_id=2b1586b7-3736-44ff-b4bd-413226eedf15, client_auth_method=client-secret

**Comment by bilias** [comment]
I've tried to revert #8734 by using the rcmail_oauth.php from 1.6.0 but nothing changed

**Comment by aleshaczech** [comment]
The issue seems to be that while in 1.6.0, the function `rcmail:url([], true, true)` rendered (in my local development environment) value https://localhost/?_task=login, in 1.6.1 the same function renders https://localhost/index.php/login/?_task=login.

**TLDR:** the solution is adding this to your config file:

```php
$config['request_path'] = 'SCRIPT_NAME';
```

(which would be a great default in 1.6.1 for backwards compatibility)

---

The OAuth2 plugin simply takes the output of `rcmail:url([], true, true)`, removes the query part of the URL and adds `/index.php/login/oauth`. Hence the result is https://localhost/index.php/login/index.php/login/oauth instead of desired https://localhost/index.php/login/oauth.

The original code to get the `$base_path` in `rcmail:url()` was as follows:

```php
$base_path = '';
if (!empty($_SERVER['REDIRECT_SCRIPT_URL'])) {
    $base_path = $_SERVER['REDIRECT_SCRIPT_URL'];
}
else if (!empty($_SERVER['SCRIPT_NAME'])) {
    $base_path = $_SERVER['SCRIPT_NAME'];
}
$base_path = preg_replace('![^/]+$!', '', $base_path);
```

In my case, that took the value of `$_SERVER['SCRIPT_NAME']` ("/index.php") and removed the file component, resulting in "/".

New code uses this:

```php
protected function get_request_path()
{
    $path = $this->config->get('request_path');

    if ($path && isset($_SERVER[$path])) {
        // HTTP headers need to come from a trusted proxy host
        if (strpos($path, 'HTTP_') === 0 && !rcube_utils::check_proxy_whitelist_ip()) {
            return '/';
        }

        $path = $_SERVER[$path];
    }
    else if (empty($path)) {
        foreach (['REQUEST_URI', 'REDIRECT_SCRIPT_URL', 'SCRIPT_NAME'] as $name) {
            if (!empty($_SERVER[$name])) {
                $path = $_SERVER[$name];
                break;
            }
        }
    }
    else {
        return rtrim($path, '/') . '/';
    }

    $path = preg_replace('/[?&].*$/', '', (string) $path);
    $path = preg_replace('![^/]+$!', '', $path);

    return rtrim($path, '/') . '/';
}
```

That means that with the default configuration (config parameter "request_path" being null), it takes the path from "REQUEST_URI", which is unfortunately very different value ("/index.php/login/oauth?_task=login&_action=oauth").

Setting the new "request_path" configuration parameter to "SCRIPT_NAME" as shown above seems to solve the problem.

**Comment by bilias** [comment]
Yes that works for my as well.

**Comment by alecpl** [comment]
Indeed, `REQUEST_URI` has been made the default in 409aee8b3c1992. And I continued to use it as default in my later code adjustments. I'll change back to `SCRIPT_NAME` as a default.

**Comment by alecpl** [comment]
Fixed.

**Comment by felixauringer** [comment]
We are using Roundcube with OIDC and would be very happy if there would be a release with this fix in the near future :)
